### PR TITLE
Support check.ignorePaths config option

### LINKED ISSUE #127: [FEATURE_REQUEST] Add `--exclude-paths` option to exclude linting files under those paths
**Description of the problem/feature request**
Our project contains a `third_party` folder containing yaml files that do not concern us, I would like an option to ignore all files under that path.

**Description of the existing behavior vs. expected behavior**
Currently this is not an option

**Additional context**
N\A

### PR DESCRIPTION
**Pull request by jumbosushi**
## Problem

We would like to exclude some directories from being checked while using the tool (fix #127)

## Solution

- Add `check.ignorePaths` configuration option that supports relative, homedir, and glob paths (didn't go with `excludePaths` since it might be confusing with the existing `exclude` option)

~This is a WIP~. Let me know if you would like to see more tests or have suggestions!

### CODE REVIEW

File: pkg/lintcontext/create_contexts.go [hunk 1]
```diff
@@ -50,6 +51,12 @@ func CreateContextsWithOptions(options Options, filesOrDirs ...string) ([]LintCo
 			continue
 		}
 
+        for _, path := range ignorePaths {
+            if strings.HasPrefix(fileOrDir, path) {
+                continue fileOrDirsLoop
+            }
+        }
```
Comment (jumbosushi) [inline pkg/lintcontext/create_contexts.go:58]: This is a naive solution to do prefix check by iterating `ignorePaths` slice. Let me know if you have suggestions for more optimized solution 🤔

File: docs/configuring-kubelinter.md [hunk 2]
```diff
@@ -46,6 +46,15 @@ checks:
   addAllBuiltIn: true
 ```
 
+## Ignore paths
+
+To ignore checks on files or direcotry under certain paths, add ignored paths to `ignorePaths`.
```
Comment (janisz) [inline docs/configuring-kubelinter.md:51]: Do we support only absolute paths?

Comment (janisz) [reply docs/configuring-kubelinter.md:51]: How about supporting globs?

Comment (jumbosushi) [reply docs/configuring-kubelinter.md:51]: Relative path is supported as well, and added glob path support. Updated doc in [03c80d1](https://github.com/stackrox/kube-linter/pull/353/commits/03c80d17f4c666035f37ec3d2e6a54fc4bb8fe7d)

File: pkg/configresolver/config_resolver_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	"testing"
+
+	"github.com/mitchellh/go-homedir"
+	"github.com/stretchr/testify/assert"
+	"golang.stackrox.io/kube-linter/pkg/config"
+)
+
+var cfg *config.Config = new(config.Config)
+
+func TestHomeIgnorePath(t *testing.T) {
```
Comment (janisz) [inline pkg/configresolver/config_resolver_test.go:15]: How about using table test for this tests?

File: pkg/configresolver/config_resolver.go [hunk 4]
```diff
@@ -55,3 +58,36 @@ func GetEnabledChecksAndValidate(cfg *config.Config, checkRegistry checkregistry
...
+				continue
+			}
+			result = expandedPath
+		} else if !filepath.IsAbs(path) {
+			absPath, err := filepath.Abs(path)
+			if err != nil {
+				errorList.AddStringf("could not expand non-absolute path: %q", absPath)
+				continue
+			}
+			result = absPath
```
Comment (janisz) [inline pkg/configresolver/config_resolver.go:81]: How about extracting inner body to a func?

```
func getIgnorePath(path string) (string, error) {
	switch {
	case path[0] == '~':
		expandedPath, err := homedir.Expand(path)
		if err != nil {
			return "", errors.Wrapf(err, "could not expand path: %q", expandedPath)
		}
		return expandedPath, nil
	case !filepath.IsAbs(path):
		absPath, err := filepath.Abs(path)
		if err != nil {
			return "", errors.Wrapf(err, "could not expand non-absolute path: %q", absPath)
		}
		return absPath, nil
	default:
		return path, nil
	}
}
```

**Review by janisz — COMMENTED** [review]
Is it still a WIP?

File: pkg/configresolver/config_resolver_test.go (hunk 3 above)
Comment (jumbosushi) [reply pkg/configresolver/config_resolver_test.go]: good idea [90986c0](https://github.com/stackrox/kube-linter/pull/353/commits/90986c0d61305c473b671a7e9d26c7af7251503b)

File: docs/configuring-kubelinter.md [hunk 5]
```diff
@@ -46,6 +46,17 @@ checks:
...
```
Comment (trumant) [inline docs/configuring-kubelinter.md:51]:
```suggestion
To ignore checks on files or directories under certain paths, add ignored paths to `ignorePaths`.
```

File: pkg/lintcontext/create_contexts.go [hunk 6]
```diff
@@ -7,6 +7,7 @@ import (
 	"sort"
 	"strings"
 
+	"github.com/bmatcuk/doublestar/v3"
```
[the remark this answers is not in the document]
Comment (jumbosushi) [reply pkg/lintcontext/create_contexts.go]: Updated to v4 in [9b3adf1](https://github.com/stackrox/kube-linter/pull/353/commits/9b3adf114a36ecc6be49a0ef8e9e6de0d00ce621). I saw v4 had [Go 1.16+ req](https://github.com/bmatcuk/doublestar/blob/5260d07/README.md#about) and I originally misread kube-linter's go.mod's go version

**Comment by jumbosushi** [comment]
just following up on this @janisz @viswajithiii 👀

**Comment by janisz** [comment]
@dhaus67 PTAL

**Review by dhaus67 — CHANGES_REQUESTED** [review]
I've tested your changes with a sample configuration as such:
```yaml
checks:
  # if doNotAutoAddDefaults is true, default checks are not automatically added.
  doNotAutoAddDefaults: false

  # addAllBuiltIn, if set, adds all built-in checks. This allows users to
  # explicitly opt-out of checks that are not relevant using Exclude.
  # Takes precedence over doNotAutoAddDefaults, if both are set.
  addAllBuiltIn: false

  # include explicitly adds checks, by name. You can reference any of the built-in checks.
  # Note that customChecks defined above are included automatically.
  include:
  - "required-label-owner"
  # exclude explicitly excludes checks, by name. exclude has the highest priority: if a check is
  # in exclude, then it is not considered, even if it is in include as well.
  exclude:
  - "privileged"
  
  ignorePaths:
  - tests/**
```

and simply ran kube-linter in the repo-root like the following:
```
kube-linter --config config.yaml lint .
```

Now I'd expect, looking at the `ignorePaths` configuration, that the `tests` folder and all its subfolders would be ignored and skipped. Essentially, I would expect a behavior similar to `.gitignore`.

However, it seems as if we are only checking whether the path is a ignore path **before** [walking the file path](https://github.com/stackrox/kube-linter/pull/353/files#diff-1c457bc54ac821de3852c0e8970c4f9ace486a385f51493586cd2935a951996aR64) and will miss removing potentially ignored directory paths afterwards.

I feel like this could lead to difficulties using the configuration as-is, as people may expect a different behavior.

What would be best is to achieve the following:
- Check whether the path itself given for linting falls under any `ignoredPath`.
- When walking the file path, check each path whether it falls under any `ignoredPath`. If that's the case, ignore the path.

**Comment by janisz** [comment]
@jumbosushi What's the status of this?

**Comment by janisz** [comment]
@jumbosushi What's the status of this?

**Review by janisz — CHANGES_REQUESTED** [review]
Please resolve conflicts

**Comment by dhaus67** [comment]
@jumbosushi any interest in finalizing this PR?

If not, I'll take this one over after some time, since I feel like this is a good configuration to have.

**Comment by jumbosushi** [comment]
Feel free to finish the feature! Not sure if i'll have time to complete it

**Comment by titanlien** [comment]
Hi @dhaus67 Did you have any update for this PR?

**Comment by titanlien** [comment]
I rebased some conflicts in go.mod then created a new PR there.
https://github.com/stackrox/kube-linter/pull/460

**Comment by janisz** [comment]
Closing as #460 was merged

**Pull request closed without merging**